system/jlib/jsocket.cpp

```diff
@@ -5081,51 +5095,95 @@ class CSocketEpollThread: public CSocketBaseThread
+        {
+            started = true;
+            ForEachItemIn(i,threads)
+            {
+                threads.item(i).start();
+            }
+        }
     }
 
     void add(ISocket *sock,unsigned mode,ISocketSelectNotify *nfy)
     {
-        CriticalBlock block(sect); // JCS->MK - are these blocks necessary? epollthread->add() uses it's own CS.
+        if ( !sock || !nfy ||
+             !(mode & (SELECTMODE_READ|SELECTMODE_WRITE|SELECTMODE_EXCEPT)) )
+            throw MakeStringException(-1,"CSocketEpollHandler::add() invalid sock or nfy or mode");
 
-        /* JCS->MK, the CSocketSelectHandler variety, checks result of thread->add and spins up another handler
-         * Shouldn't epoll version do the same?
-         */
-        if (!epollthread->add(sock,mode,nfy))
-            throw MakeStringException(-1, "CSocketEpollHandler: failed to add socket to epollthread handler: sock # = %d", sock->OShandle());
+        CriticalBlock block(sect);
+        // Create new handler thread if current one has hdlPerThrd fds.
+        // epoll() handles many fds faster than select so this would
+        // seem not as important, but we are still serializing on
+        // nfy events and spreading those over threads may help,
+        // especially with SSL as avail_read() could block more.
+        for (;;)
+        {
+            bool added=false;
+            ForEachItemIn(i,threads)
+            {
+                if (added)
+                {
+                    if (threads.item(i).remove(sock))
+                        return;
+                }
+                else
+                    added = threads.item(i).add(sock,mode,nfy);
+            }
+            if (added)
+                return;
+            CSocketEpollThread *thread = new CSocketEpollThread(epolltrace, hdlPerThrd);
+            threads.append(*thread);
+            if (started)
+                thread->start();
+        }
     }
 
     void remove(ISocket *sock)
     {
-        CriticalBlock block(sect); // JCS->MK - are these blocks necessary? epollthread->add() uses it's own CS.
-        epollthread->remove(sock);
+        CriticalBlock block(sect);
+        ForEachItemIn(i,threads)
+        {
+            if (threads.item(i).remove(sock))
+                break;
+        }
     }
 
     void stop(bool wait)
     {
-        IException *e=NULL;
-        epollthread->stop(wait);           // not quite as quick as could be if wait true
-        if (wait && !e && epollthread->termexcept)
-            e = epollthread->termexcept.getClear();
+        IException *e = nullptr;
+        CriticalBlock block(sect);
+        ForEachItemIn(i,threads)
+        {
+            CSocketEpollThread &t=threads.item(i);
+            {
+                CriticalUnblock unblock(sect);
+                t.stop(wait);           // not quite as quick as could be if wait true
+            }
+            if (wait && !e && t.termexcept)
+                e = t.termexcept.getClear();
```

Comment:
picky: Could potentially have multiple exceptions and this might leak one.  It is never thrown, so may be worth deleting this code completely (with a comment)